temporalio/contrib/pydantic.py

```diff
@@ -0,0 +1,120 @@
+
+import temporalio.workflow
+from temporalio.converter import (
+    AdvancedJSONEncoder,
+    CompositePayloadConverter,
+    DataConverter,
+    DefaultPayloadConverter,
+    JSONPlainPayloadConverter,
+    JSONTypeConverter,
+)
+from temporalio.worker.workflow_sandbox._restrictions import RestrictionContext
+
+# Note that in addition to the implementation in this module, _RestrictedProxy
+# implements __get_pydantic_core_schema__ so that pydantic unwraps proxied types
+# when determining the schema.
+
+
+class PydanticModelTypeConverter(JSONTypeConverter):
+    """Type converter for pydantic model instances."""
+
+    def to_typed_value(self, hint: Type, value: Any) -> Any:
+        """Convert dict value to pydantic model instance of the specified type"""
+        if not inspect.isclass(hint) or not issubclass(hint, pydantic.BaseModel):
+            return JSONTypeConverter.Unhandled
+        model = hint
+        if not isinstance(value, dict):
+            raise TypeError(
+                f"Cannot convert to {model}, value is {type(value)} not dict"
+            )
+        if temporalio.workflow.unsafe.in_sandbox():
+            # Unwrap proxied model field types so that Pydantic can call their constructors
+            model = pydantic.create_model(
+                model.__name__,
+                **{  # type: ignore
+                    name: (RestrictionContext.unwrap_if_proxied(f.annotation), f)
+                    for name, f in model.model_fields.items()
+                },
+            )
+        if hasattr(model, "model_validate"):
+            return model.model_validate(value)
+        elif hasattr(model, "parse_obj"):
+            # pydantic v1
+            return model.parse_obj(value)
+        else:
+            raise ValueError(
+                f"{model} is a Pydantic model but does not have a `model_validate` or `parse_obj` method"
+            )
+
+
+class PydanticJSONEncoder(AdvancedJSONEncoder):
+    """JSON encoder for python objects containing pydantic model instances."""
+
+    def default(self, o: Any) -> Any:
+        """Convert object to jsonable python.
+
+        See :py:meth:`json.JSONEncoder.default`.
+        """
+        if isinstance(o, pydantic.BaseModel):
+            return to_jsonable_python(o)
+        return super().default(o)
+
+
+class PydanticPayloadConverter(CompositePayloadConverter):
+    """Payload converter for payloads containing pydantic model instances.
+
+    JSON conversion is replaced with a converter that uses
+    :py:class:`PydanticJSONEncoder` to convert the python object to JSON, and
+    :py:class:`PydanticModelTypeConverter` to convert raw python values to
+    pydantic model instances.
+    """
+
+    def __init__(self) -> None:
+        """Initialize object"""
+        json_payload_converter = JSONPlainPayloadConverter(
+            encoder=PydanticJSONEncoder,
+            custom_type_converters=[PydanticModelTypeConverter()],
+        )
```

Comment:
Our pydantic sample was implementing `to_payload`. This is slightly incorrect because it means that our `AdvancedJSONEncoder` was not being used. In practice, pydantic provides a function (called `to_jsonable_python` in v2) with similar functionality to our `AdvancedJSONEncoder`, but nevertheless, we should ensure that we fall back to `AdvancedJSONEncoder` in case we add to it in the future.